WIP: 型安全性を高めるためのzod導入とリファクタリング進捗報告とご相談

kei615ykhm opened the pull request:
このプルリクエストはマージを目的としていないため、セッション終了後にクローズする予定です。

## 概要
作業途中のため、以下の項目が主な内容になります。

- 作業進捗の報告
- 現時点のコードが型安全を維持できているかの確認
- スキーマ定義に関するご相談

現在、Issue: #16 に基づき、PR: #14 以降の内容から `zod` を導入し、バリデーションの実装と型安全性向上を目的としたリファクタリングを進めています。

なお、#17 にてESLint、Prettier、tsconfigルール設定等の環境調整を行った後に作成したブランチでの作業になります。

## 作業状況
1. `src/schemas` ディレクトリの作成
   - `src/schemas/memoSchema.ts`: メモのスキーマを定義

2. `src/types` ディレクトリの更新
   - `src/types/index.ts`: zod スキーマから型を生成

3. `src/utils` ディレクトリの作成
   - `src/utils/zodParser.ts`: 汎用的な zod パース関数を実装

4. `src/hooks/useMemoManager.ts` の更新
   - zod スキーマを使用してバリデーションを実装
   - 新規メモ作成時にバリデーションを適用
   - ローカルストレージへの保存機能を追加

5. `src/types` ディレクトリの削除
   - 型定義が `src/schemas/memoSchema.ts` と重複するため

6. `src/schemas/memoSchema.ts` で定義したメモスキーマの改善（作業中）
   - commit: 176f003 〜 dd58504 にて作業

※作業途中で得た知見を基に、一部の予定のタスクを変更している箇所もあります。

## ご相談内容
`src/hooks/useMemoManager.ts` にて zod のバリデーションを実装した際に学んだ内容を基に、`src/schemas/memoSchema.ts` に定義しているメモスキーマにバリデーションを追加する作業を進めています。

```ts
import { z } from 'zod';

export const memoSchema = z.object({
  id: z.string(),
  content: z.string(),
  createdAt: z.string(),
});

export type Memo = z.infer<typeof memoSchema>;
```
現時点で、`createdAt`のバリデーション強化、`content`の制約追加、`id`の一意性保証を行いました。しかし、エラーメッセージの詳細化について理解が不十分です。現在、バリデーションエラーメッセージが具体的でないという問題を認識していますが、それぞれのエラーメッセージにどのような内容を設定すべきかがわかりません。どのようなエラーメッセージを設定すべきか、ご助言いただければ幸いです。

kei615ykhm committed:
add: `zod`ライブラリでメモオブジェクトのスキーマを定義

kei615ykhm committed:
add: `Memo`型の再エクスポート

- 他のファイルから`src/types`経由でアクセス可能にする

kei615ykhm committed:
add: `memoSchema`スキーマから推論された`Memo`型のエクスポート

kei615ykhm committed:
remove: 二重定義している`Memo`型を削除し、`memoSchema`で統一する

kei615ykhm committed:
docs: TODOコメントでスキーマで定義した型への変更を説明

kei615ykhm committed:
change: `zod`スキーマで定義した`Memo`型を使用するためにパスを変更

kei615ykhm committed:
add: データを安全にパースする関数を定義

kei615ykhm committed:
docs: TSDocコメントでパース関数を説明

kei615ykhm committed:
docs: 実装を終えたため`zod`のスキーマ定義を示すコメントを削除

kei615ykhm committed:
change: `memoSchema`を使用してメモの配列スキーマを定義

kei615ykhm committed:
add: パースしたメモをバリデートする

- バリデーションが成功した場合は`validatedMemos`にバリデート済みのメモが格納される

kei615ykhm committed:
update: `memoSchema`と`zod`を宣言

kei615ykhm committed:
docs: 実装した機能のTODOを削除

kei615ykhm committed:
change: 新しいメモをバリデーションする

kei615ykhm committed:
docs: 実装した機能のTODOを削除

kei615ykhm committed:
style: 行間の調整

kei615ykhm committed:
docs: TSDocコメントで新規メモをバリデーションする説明を追加

kei615ykhm committed:
remove: `Memo`型の定義を`memoSchema`に移行したため、削除

kei615ykhm committed:
update: 曖昧な命名をしている`Memo`型のオブジェクトを明示的な名前に変更

kei615ykhm committed:
docs: NOTEコメントでバリデーションをスキップしていることを説明

- まだ保存機能を実装していないため

kei615ykhm committed:
update: `Memo`型を明示的に指定し暗黙的な`any`への認識を禁止

kei615ykhm committed:
update: `JSON.parse`の戻り値を`unknown`型として扱う

kei615ykhm committed:
update: バリデーション結果を状態として設定

- `parsedMemos`が`Memo[]`型であると保証される

kei615ykhm committed:
remove: 不要になったエラーハンドリングの分岐コードを削除

kei615ykhm committed:
docs: 書き換えたことで不要になったNOTEコメントを削除

kei615ykhm committed:
update: `setMemos`の更新を関数型で行うように変更

- 最新の状態を基にメモを追加できるようにする

kei615ykhm committed:
add: ローカルストレージへの保存機能を追加

kei615ykhm committed:
add: メモを削除したときの情報をローカルストレージに保存できるように設定

kei615ykhm committed:
remove: 重複したコードの削除

kei615ykhm committed:
update: 結果を`updatedMemos`変数に格納し、更新されたメモのリストを明確に示す

- `49b5ac63bcfbf2fb46741fa97733f64b873ed622`での作業に引き続き可読性を向上させるアプローチ

kei615ykhm committed:
docs: アノテーションコメントで現状の問題点を把握・整理する

kei615ykhm committed:
add: 有効なISO日付形式であることを確認するためのカスタムバリデーションを追加

kei615ykhm committed:
docs: NOTEコメントでコードを補助的に説明

kei615ykhm committed:
docs: 改善済みのWARNINGコメントを削除

kei615ykhm committed:
add: `content`フィールドに対して最低`1`文字以上を要求するバリデーションを追加

kei615ykhm committed:
docs: 改善済みのWARNINGコメントを削除

kei615ykhm committed:
add: 重複を防止する対策で`UUID`を`id`として使用

kei615ykhm committed:
docs: 改善済みのWARNINGコメントを削除

kei615ykhm committed:
docs: 自己説明的なコメントを削除

kei615ykhm committed:
add: `strict`メソッドを使用して定義してないフィールドが含まれる場合にバリデーションエラーを発生させるようにする

kei615ykhm committed:
docs: 改善済みのWARNINGコメントを削除

niaka3dayo commented:
> 
> ```ts
> import { z } from 'zod';
> 
> export const memoSchema = z.object({
>   id: z.string(),
>   content: z.string(),
>   createdAt: z.string(),
> });
> 
> export type Memo = z.infer<typeof memoSchema>;
> ```
> 
> 現時点で、`createdAt`のバリデーション強化、`content`の制約追加、`id`の一意性保証を行いました。しかし、エラーメッセージの詳細化について理解が不十分です。現在、バリデーションエラーメッセージが具体的でないという問題を認識していますが、それぞれのエラーメッセージにどのような内容を設定すべきかがわかりません。どのようなエラーメッセージを設定すべきか、ご助言いただければ幸いです。

`memoSchema`および`Memo`のデータベースでの要件が定まっていないため、現時点でそれぞれの値がどのような形をとっているのが正しいのかというのが見えておらず、そもそも「どういう状況が正しいのか」という要件が存在しないのではないでしょうか？
なぜなら、この`Memo`という型は、本来はデータベースのテーブルの1レコードあたりの形と制限に依存するからです。

なので、たとえば、以下のようにデータベースのスキーマをPrismaで定義する予定だという仮定にしてみます。

```prisma
// prisma/schema.prisma
model Memo {
  id        String   @id @default(uuid())
  content   String   @db.Text
  createdAt DateTime @default(now())
}
```


すると、データベースからFetchしたとき、かえってくるべきデータは以下の要件となります。

### id:

- 型: UUID形式の文字列。
- 必須: 必ず存在する必要があります。

### content:

- 型: 文字列。
- 必須: 必ず存在する必要があります。
- 制約: 最小文字数: 1文字以上。 最大文字数: 1000文字以内（データベースの@db.Textに基づく適切な長さ）。

### createdAt:

- 型: ISO 8601形式の日時文字列。
- 必須: 必ず存在する必要があります。

この要件を満たすためのエラーメッセージ入りの`memoSchema`は以下の通りです。

```ts
export const memoSchema = z.object({
  id: z
    .string({ message: 'IDは文字列である必要があります。' })
    .uuid({ message: 'IDは有効なUUID形式でなければなりません。' }),
  
  content: z
    .string( { messsage: '内容は文字列で定義する必要があります。'} )
    .min(1, { message: '内容は最低1文字必要です。' })
    .max(1000, { message: '内容は1000文字以内でなければなりません。' }),
  
  createdAt: z
    .string( {messsage: '日付は有効な日付形式文字列で定義する必要があります。'} )
    .refine((val) => !isNaN(Date.parse(val)), { message: '作成日時は有効な日付形式でなければなりません。' }),
});
```

また、データを取得したものが正しいかを判定するのに使用する`zodSchema`とフォームがデータを作成する際にバリデーションを行うためのスキーマは、型も目的も違うので別途定義する必要があります。

フォームから新しいメモを作成する際には、idやcreatedAtはクライアント側で生成する必要がないため、これらのフィールドを除外したスキーマを定義します。

```ts
import { z } from 'zod';

/** メモ作成用のスキーマ */
export const createMemoSchema = z.object({
  content: z
    .string({ message: '内容は文字列で定義する必要があります。' })
    .min(1, { message: '内容は最低1文字必要です。' })
    .max(1000, { message: '内容は1000文字以内でなければなりません。' }),
});

/** メモ作成時の型 */
export type CreateMemoInput = z.infer<typeof createMemoSchema>;
```


こういう相談内容はあとから見返せしやすいように、PRで行うより本来はIssueで残したほうがいいかもしれないです。

niaka3dayo reviewed (commented):
いただいていた相談に対応しました

kei615ykhm commented:
@niaka3dayo ご対応ありがとうございます。
確かに、PRよりもissueとして対応する方が適切でしたね。
本件について再度確認したい点がありますので、後ほどissueを作成し、改めてアサインさせていただきます。

また、別件でお願いしていた件についてですが、私の書き方が悪く、うまく伝わっていなかったようで申し訳ありません。

改めてになりますが、各ファイル内のコードの型安全性についても確認をお願いしたいと考えておりました。
相談内容へのご返答で「データを取得する際に使用するzodSchemaと、フォームでデータを作成する際のバリデーション用スキーマは、型や目的が異なるため、別途定義する必要がある」とのことでしたが、これが現在のコードの問題点でもあるという理解で間違いないでしょうか？

niaka3dayo commented on src/schemas/memoSchema.ts:
```diff
@@ -0,0 +1,13 @@
+
+export const memoSchema = z
+  .object({
+    id: z.string().uuid({ message: 'IDは有効なUUIDである必要があります' }),
```
[SHOULD] `uuid`の意味をわかって使っていなさそうです。現時点ではローカルストレージをデータベースのように使用しているだけなので私の出した例をコピペしたのでは意味が違います。自分で設計したIDの形式でスキーマを直してください。

kei615ykhm replied:
@niaka3dayo 
ご指摘いただいた通り、`uuid`のことを全く理解しないまま使っていたと気づいたので再学習して書き直してきました。こういう感じでしょうか？

コミット: abe004f363f4fc56c1ed2ae1f08150e75cf0b4df

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -48,22 +36,35 @@ export const useMemoManager = () => {
   const handleAddMemo = (content: string) => {
-    const newMemo: Memo = {
+    const newMemoData: Memo = {
       id: Date.now().toString(),
```
[SHOULD] IDに単純な日付文字列を使用するのは危険です。全く同時に別のウィンドウでメモを追加したりしたら簡単にバグになります。きちんと一意なID算出方法を設計したほうがいいです。それこそUUIDとか。

kei615ykhm replied:
@niaka3dayo
`memoSchema.ts`で`uuid`を定義しておきながら`id: Date.naw().toString(),`という書き方では一意性の保証が不十分だったということでしょうか。

コミット: adc104a20ca01ff62309c67cf4dbbe927b42dc3d, fb71f784307d43a97deed1cf98b9ec131018fe38

補足: コミットメッセージが説明不十分でした。名前衝突の回避や一貫性の観点からもエイリアス使用のほうが良いのかなと考えて書いています。

niaka3dayo commented on src/hooks/useMemoManager.ts:
```diff
@@ -48,22 +36,35 @@ export const useMemoManager = () => {
    */
   const handleDeleteMemo = (id: string) => {
-    setMemos(memos.filter((memo) => memo.id !== id));
+    const updatedMemos = memos.filter((memo) => memo.id !== id);
+    setMemos(updatedMemos);
+    /** ローカルストレージに削除情報を保存 */
+    localStorage.setItem(
+      'memos',
+      JSON.stringify(memos.filter((memo) => memo.id !== id))
+    );
```
[SHOULD] `filter`みたいな重たいループメソッドを全く同じ内容でわざわざ2回呼んでいるのはパフォーマンスに影響が出そうです。
`updatedMemos`を流用してローカルストレージに保存すればいいと思います。

kei615ykhm replied:
@niaka3dayo 
こういうことでしょうか？

コミット: f548527524aabdbe2db87f3d2a1191e3a8fae46e

niaka3dayo commented on src/utils/zodParser.ts:
```diff
@@ -0,0 +1,12 @@
+ * @param data - バリデーション対象のデータ
+ * @returns バリデーションが成功した場合はデータ、失敗した場合はnull
+ */
+export function safeParse<T>(schema: z.ZodType<T>, data: unknown): T | null {
```
[SHOULD] `Zod`スキーマの`safeParse`とほぼやっていることが同じなのでかなり冗長なわりに機能のない関数です。
どうせならドメインモデルの型ごとに関数を作ってあげたいし、このままじゃ型ガードの役割もできないかなと思います。

kei615ykhm replied:
@niaka3dayo いただいたアドバイスの意味を理解するのに時間がかかってしまいました。
先日、教えて頂いたドメイン駆動設計のお話を反映したものではありませんが、ドメインモデルの型ごとに関数を作るというのは、こういうことでしょうか？

補足: 試行錯誤でコミットが複数になってしまったのでハッシュにまとめられませんでした。

```ts
/**
 * 未知のデータをMemo型に変換する
 * @param data 検証するデータ
 * @returns データが有効な場合はMemoオブジェクト、そうでない場合はnull
 */
export function parseMemo(data: unknown): Memo | null {
  const result = memoSchema.safeParse(data);
  if (result.success) {
    return result.data;
  } else {
    console.error('Memo parsing failed:', result.error);
    return null;
  }
}

/**
 * 新しいメモ入力をバリデーションする
 * @param data 検証するデータ
 * @returns データが有効な場合はCreateMemoInputオブジェクト、そうでない場合はnull
 */
export function validateCreateMemoInput(data: unknown): CreateMemoInput | null {
  const result = createMemoSchema.safeParse(data);
  return result.success ? result.data : null;
}

/**
 * データがMemo型かどうかを判定する型ガード関数
 * @param data 検証するデータ
 * @returns dataがMemo型の場合true、そうでない場合false
 */
export function isMemo(data: unknown): data is Memo {
  return memoSchema.safeParse(data).success;
}
```

niaka3dayo reviewed (changes requested):
@kei615ykhm レビューしました。

kei615ykhm committed:
chore: `uuid`ライブラリのインストール

kei615ykhm committed:
chore: `moduleResolution`の設定を`node`に変更

kei615ykhm committed:
chore: `allowSyntheticDefaultImports`の追加と有効化

kei615ykhm committed:
chore: 依存関係の再インストール

kei615ykhm committed:
chore: `@types/uuid`のインストール

- `node_modules/uuid/dist/index.js` が暗黙的に`any`型になる問題の解決を図るために実施

kei615ykhm committed:
update: `uuid`から`v4`関数をインポートしてUUIDが正しく生成できるように修正

kei615ykhm committed:
update: `v4`にエイリアスを付けて可読性の向上を図る

kei615ykhm committed:
chore: `jest`のインストール

kei615ykhm committed:
add: jestの設定ファイルを作成

kei615ykhm committed:
add: jestのスクリプトを追加

kei615ykhm committed:
update: `globals`セクションの削除

- 前回のテスト時に非推奨設定であると警告されたため

kei615ykhm committed:
disable: `setupFilesAfterEnv`オプションを無効化

kei615ykhm committed:
update: メモのバリデーションを追加

- 各フィールドに対してバリデーションを追加
  - id: UUID形式であることをバリデーション
  - content: 最低1文字、最大1000文字の制約を追加
  - createdAt: 有効な日付形式であることをバリデーション

kei615ykhm commented:
@niaka3dayo 
返答と修正が遅れてしまい申し訳ございません。全体的に知識不足と言語化できていない部分があったため、`ID`と`UUID`の周辺知識を再学習しながら作業しています。

[zennスクラップ](https://zenn.dev/kei615ykhm/scraps/8b78167130ee81)

修正が済み次第、返信させていただきますので、よろしくお願い致します。

kei615ykhm committed:
perf: `updatedMemos`変数を再利用して`filter`メソッドの呼び出しを1回にする

- `memos.filter((memo) => memo.id !== id`の結果を`updatedMemos`に格納し、その変数を`setMemos`と`localStorage.setItem`の両方で再利用

kei615ykhm committed:
disable: パース関数を無効化

- ドメインモデルに適した形式の関数を新規で作成するために一時期的に無効化する。作成が済み次第、削除予定

kei615ykhm committed:
docs: TODOコメントとして関数作成後にファイルを削除すると追記

kei615ykhm committed:
add: メモ作成用のスキーマを定義

kei615ykhm committed:
remove: 重複しているバリデーションを削除

kei615ykhm committed:
docs: スキーマの説明

kei615ykhm committed:
add: 未知のデータを`Memo`型に変換する関数

kei615ykhm committed:
add: 新しいメモ入力をバリデーションする関数

kei615ykhm committed:
update: 型の命名を変更

- `CreateMemo`から`CreateMemoInput`に変更

kei615ykhm committed:
remove: スキーマを無視して再定義している`createMemoSchema`を削除

kei615ykhm committed:
add: `createMemoSchema`をインポート

kei615ykhm committed:
refactor: 型とスキーマをグループ化してインポート

kei615ykhm committed:
remove: パース関数の移行が完了したため削除

kei615ykhm committed:
add: 型ガード関数の定義

kei615ykhm commented:
Issue: #20 タスクとして、このプルリクエストに適したマイルストーンを設定し直しました。

kei615ykhm commented:
@niaka3dayo 修正が遅れてしまい申し訳ありません。

ここまでにご指摘、アドバイスいただいた項目の修正は完了しましたので、各項目にてご確認をお願いいたします。

ドメインモデルの概念や目的については概ね把握しました。しかし、その使い方と設定理論については今後の学習課題です。
また、作業する中で名前の衝突と拡張性についても課題を感じました。そのため、先日教えて頂いたドメイン駆動設計と`package by feature`を組み合わせた設計パターンを採用し、以下のように作ってみたいと考えています。

```scss
src/
├── domains/
│   └── memo/
│       ├── memoParser.ts         // メモのパース処理
│       ├── memoValidator.ts      // ランタイムバリデーション
│       ├── memoTypeGuard.ts      // 型ガード
│       ├── memoZodSchema.ts      // Zodスキーマ定義
│       ├── memoDataRepository.ts // リポジトリ層
│       └── memoApplicationService.ts // アプリケーションサービス
└── features/
    └── memo/
        └── useMemoStateManager.ts // メモの状態管理
```

### より詳細な情報
Issue: [`memoUtils.ts`の責任範囲分割とファイル移行計画の提案](https://github.com/kei615ykhm/logic-loom-nextjs14/issues/22)

こちらの移行作業は、今回のプルリクエストの主旨と異なる内容なので、一旦実験用ブランチにて作業してみます。後ほど、別件としてコードレビュー依頼をさせてください。
